normalized-cache-incubating/src/commonMain/kotlin/com/apollographql/cache/normalized/internal/DefaultApolloStore.kt

```diff
@@ -118,11 +192,111 @@ internal class DefaultApolloStore(
         cacheKey = CacheKey.rootKey(),
         variables = variables,
         fieldKeyGenerator = fieldKeyGenerator,
+        returnPartialResponses = true,
     )
-    return ReadResult(
-        data = batchReaderData.toData(operation.adapter(), customScalarAdapters, variables),
-        cacheHeaders = batchReaderData.cacheHeaders,
-    )
+    val dataWithErrors: Map<String, Any?> = batchReaderData.toMap()
+    val errors = mutableListOf<Error>()
+
+    @Suppress("UNCHECKED_CAST")
+    val dataWithNullBubbling: Map<String, Any?>? = nullBubble(dataWithErrors, operation.rootField(), errors) as Map<String, Any?>?
+    val falseVariablesCustomScalarAdapter =
+      customScalarAdapters.newBuilder().falseVariables(variables.valueMap.filter { it.value == false }.keys).build()
+    val data = dataWithNullBubbling?.let { operation.adapter().fromJson(it.jsonReader(), falseVariablesCustomScalarAdapter) }
+    return ApolloResponse.Builder(operation, uuid4())
+        .data(data)
+        .errors(errors.takeIf { it.isNotEmpty() })
+        .cacheHeaders(batchReaderData.cacheHeaders)
+        .cacheInfo(
+            CacheInfo.Builder()
+                .fromCache(true)
+                .cacheHit(errors.isEmpty())
+                .stale(batchReaderData.cacheHeaders.headerValue(ApolloCacheHeaders.STALE) == "true")
+                .build()
+        )
+        .build()
+  }
+
+  /**
+   * If a position contains an Error, replace it by a null if the field's type is nullable, bubble up if not.
+   */
+  private fun nullBubble(dataWithErrors: Any?, field: CompiledField, errors: MutableList<Error>): Any? {
```

Comment:
Nit: the spec doesn't mention "bubbles", only "error propagation"

```suggestion
  /**
   * If a position contains an Error, replace it by a null if the field's type is nullable, propagage the error if not.
   */
  private fun propagateErrors(dataWithErrors: Any?, field: CompiledField, errors: MutableList<Error>): Any? {
```